src/util/SearchUtility.test.js

```diff
@@ -281,3 +290,37 @@ test("SearchUtility should support custom tokenizer pattern", async done => {
   expect(await searchUtility.search("6o")).toEqual([6]);
   done();
 });
+test("SearchUtility should use maxSubstringLength for ALL_SUBSTRINGS index mode", async done => {
+  const searchUtility = init({
+    indexMode: INDEX_MODES.ALL_SUBSTRINGS,
+    maxSubstringLength: 10
+  });
+  expect(
+    (await searchUtility.search("Verylongstringwithoutdelimiter")).length
+  ).toBe(0);
+  expect(await searchUtility.search("verylong")).toEqual([8]);
+  expect(await searchUtility.search("without")).toEqual([8]);
+  expect(await searchUtility.search("delimiter")).toEqual([8]);
+  expect((await searchUtility.search("withoutdelimiter")).length).toBe(0);
```

Comment:
I think this change is better... I'm still unsure about how `maxSubstringLength` affects all substrings mode though.

Given a token "Verylongstringwithoutdelimiter ", what's the value of being able to search and match on e.g. "without" and "delimiter" rather than just "verylong"?

I think I would have expected the indexer to take a word like "Verylongstringwithoutdelimiter", see there's a max subtring length of 8, truncate it to "Verylong" and index that only (discarding the rest– because the rest is arbitrary).

Reply:
What this is essentially doing is putting an upper limit on the input search value.
Internally in our product, a user might have some input like:

"1000-3000-1400-11100...." (assume very long input)

We caused a sev by trying to calculate all substrings of this. With a max substring length of 10, all the individual numbers here would still be searchable.

Reply:
Right. Just to be clear, I understand the memory usage concerns of trying to pre-calculate all substring indexes for a document with longer strings. So I understand the motivation for putting some kind of cap on that.

My concern is that the way the cap is implemented, the resulting search becomes difficult to predict. For example, if we had three documents that contained the following strings:
* A: cat
* B: concatenate
* C: sophisticated

And we passed a `maxSubstringLength` of 3. You'd be able to match A and B with the search string "cat" but not C because "sophisticated" would get broken down into separate chunks to be indexed: "sop", "his", "tic", "ate", "d". This seems too arbitrary to me to be a good user experience. So if you typed "t" and saw all three matches, then refined to "at" and saw all three, then refined to "cat" and now only see two– that's an expected behavior I think. At least if we only indexed the first `maxSubstringLength` characters then finding A would seem more consistent.

I wonder if what you really want here is the ability to specify a custom tokenizer so that you can split your example string above ("1000-3000-1400-11100....") into more meaningful chunks of "1000", "3000", "1400", "11100", etc. (which might not all even be the same length).

Reply:
I see your confusion. This implementation is not changing the tokens, so 'sophisticated' gets passed into SearchIndex. 
SearchIndex is actually breaking down sophisticated into:
`[s, so, sop, o, op, oph, h, hi, his, ....c, ca, cat, a, at, ate, ....ted]` e.g. all substrings less than `maxSubstringLength`

So 'cat' would actually match 'sophisticated'. 

Agree about the custom tokenizer, but this is also a fallback because our users can enter any values.

Reply:
Gotcha.

Hm. I need to think about this some more.

I understand your use case but I'm not sure if I want to take on supporting this workaround.